---
title: "R Notebook"
output: html_notebook
---

The project regards the discovery of disease subtypes using a multi-omics dataset coming from TCGA. The dataset is the Prostate adenocarcinoma dataset (disease code “PRAD”). We will consider as disease subtypes the ones identified in a work performed by The Cancer Genome Atlas Research Network [1], where they used an integrative clustering model (called iCluster [2]) on multi-omics data (somatic copy-number alterations, methylation, mRNA, microRNA, and protein levels) and discovered three disease subtypes.

1.  Download the Prostate adenocarcinoma dataset considering three different omics data sources (mRNA, miRNA and protein expression data). The TCGA code for the dataset is “PRAD”.

```{r}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("curatedTCGAData");
BiocManager::install("TCGAbiolinks");
BiocManager::install("TCGAutils");

install.packages("SNFtool");
install.packages("NetPreProc");
install.packages("caret");
install.packages("cluster");
install.packages("mclustcomp");
```

```{r}
library("curatedTCGAData")
library("TCGAbiolinks");
library("TCGAutils");
library("NetPreProc");
library("caret");
library("cluster"); #pam
library("SNFtool");
library("mclustcomp");
```

```{r}
#assay refers to a specific type of experimental measurement or analysis
#performed on a biological sample to capture data at a particular molecular #level. Each assay represents a distinct omics layer, such as:
#Genomics Assay, Transcriptomics Assay, Proteomics Assay, Metabolomics Assay

assays <- c("RNASeq2Gene", "miRNASeqGene", "RPPAArray");
multi_omics_dataset <- curatedTCGAData(diseaseCode = "PRAD", 
                        assays = assays, 
                        version = "2.0.1", dry.run = FALSE);
multi_omics_dataset
```

2.  Pre-process the dataset following the same steps we used during lessons. During the filtering by variance, select the first 100 features having highest variance from each data source.

```{r message=FALSE, warning=FALSE}
# Consider only primary solid tumors:
primary <- TCGAutils::TCGAsampleSelect(colnames(multi_omics_dataset), c("01"));
multi_omics_dataset <- multi_omics_dataset[, primary, ];

# Check for replicates (anyReplicated checks the so called biological or 
# primary unit in the sampleMap of the MultiAssayExperiment object, that
# corresponds to the first 12 characters of the barcodes for TCGA data):
check_rep <- anyReplicated(multi_omics_dataset);
print(check_rep);

# The information regarding if the sample is FFPE is stored in the clinical data,
# which are accessible using colData(). 
no_ffpe <- which(as.data.frame(colData(multi_omics_dataset))$patient.samples.sample.is_ffpe == "no");
multi_omics_dataset <- multi_omics_dataset[, no_ffpe, ];

# Obtain samples having all the considered omics:
complete <- intersectColumns(multi_omics_dataset);

# Extract assays in list:
complete <- assays(complete);

# Obtain matrices samples x features:
#Transposes the matrix
complete <- lapply(complete, FUN=t)

# Remove features having NAs (present only in proteomics data):
complete[[3]] <- complete[[3]][, colSums(is.na(complete[[3]])) == 0];

# Remove features with near zero variance and retain top 500 features 
# having higher variance:
features_threshold <- 100;
for(i in 1:length(complete)){
    zero_var_indexes <- caret::nearZeroVar(complete[[i]])
    message(paste("Removed ", length(zero_var_indexes), "features from", names(complete)[i]));
    
    if(length(zero_var_indexes) != 0){
        complete[[i]] <- complete[[i]][, -zero_var_indexes];
    }

    # if less features than threashold return them all
    if(ncol(complete[[i]]) <= features_threshold) next
    
    vars <- apply(complete[[i]], 2, var);
    var_ordered_indexes <- sort(vars, index.return=TRUE, decreasing = TRUE)$ix;
    
    complete[[i]] <- complete[[i]][, var_ordered_indexes[1:features_threshold]];
}

# Perform features standardization using z-score:
zscore <- function(data){
    zscore_vec <- function(x) { return ((x - mean(x)) / sd(x))}
    data <- apply(data, 2, zscore_vec)
    data
}

complete <- lapply(complete, zscore);

# Clean barcodes retaining only "Project-TSS-Participant":
for(v in 1:length(complete)){
    rownames(complete[[v]]) <- substr(rownames(complete[[v]]), 1, 12);
}
```

3.  Download the disease subtypes (column “Subtype_Integrative” is the one containing the iCluster molecular subtypes). Note that not all subtypes are available for the set of samples having all the considered omics data sources, thus you need to retain from the multi-omics dataset only samples having an associated subtype.

```{r}
# Download disease subtypes from TCGAbiolinks:
subtypes <- as.data.frame(TCGAbiolinks::PanCancerAtlas_subtypes());
subtypes <- subtypes[subtypes$cancer.type == "PRAD", ];

# Retain only primary solid tumors and select samples form multi-omics dataset having an associated subtype
# (in the same order):
subtypes <- subtypes[TCGAutils::TCGAsampleSelect(subtypes$pan.samplesID, "01"), ];
samples_with_subtype <- rownames(complete[[1]]) %in% substr(subtypes$pan.samplesID, 1, 12);
subtypes_with_sample <- substr(subtypes$pan.samplesID, 1, 12) %in% rownames(complete[[1]])

subtypes <- subtypes[subtypes_with_sample,]
rownames(subtypes) <- substr(subtypes$pan.samplesID, 1, 12);

for(v in 1:length(complete)) {
  complete[[v]] <- complete[[v]][samples_with_subtype,];
}
```

4.  Check that patients in multi-omics dataset and subtypes are in the same order.

```{r}
subtypes <- subtypes[order(subtypes$pan.samplesID), ]
for(v in 1:length(complete)) {
  complete[[v]] <- complete[[v]][order(rownames(complete[[v]])),];  
}
all(rownames(subtypes) == rownames(complete[[1]]))
```

5.  Integrate the data using Similarity Network Fusion [3] with the scaled exponential euclidean distance.

```{r}
# Compute similarity matrix for each data source using the scaled
# exponential euclidean distance:
W_list <- list();
for(i in 1:length(complete))
{
    distances_matrix <- (dist2(as.matrix(complete[[i]]), as.matrix(complete[[i]])))^(1/2);
    W_list[[i]] <- affinityMatrix(distances_matrix);
}
    
# Integration of multi-omics data using Similarity Network Fusion:
# t is the number of iterations and K is the number of neighbors to 
# consider to compute the local similarity matrix:
W_int <- SNF(W_list, K=20, t=20);
```

6.  Try to integrate the similarity matrices from each data source (computed by scaled exponential euclidean distance) using a simple average of the matrices. This can be considered as a trivial multi-omics data integration strategy.

```{r}
W_averages <- Reduce(`+`, W_list) / length(W_list)
```

7.  [GROUP] Integrate the dataset using another data fusion method called NEMO [4] to obtain an integrated similarity matrix. NEMO implementation is available on github (<https://github.com/ShamirLab/NEMO>).

```{r}

```

8.  Perform disease subtype discovery (number of clusters equal to the number of disease subtypes found by iCluster) using PAM algorithm [5] on the following similarity matrices:

a.  Similarity matrices obtained from single data sources (i.e. miRNA, mRNA, proteins) using the usual scaled exponential euclidean distance. Thus, you should obtain three different similarity matrices. To compute the corresponding distance matrix use this code: dist \<- 1 - NetPreProc::Max.Min.norm(W). Max.Min.norm() function is in the NetPreProc CRAN package (<https://cran.r-project.org/web/packages/NetPreProc/index.html>). The idea is to normalize the similarity matrix before computing the corresponding distance.
b.  Integrated matrix obtained using the average among matrices. Use dist \<- 1 - NetPreProc::Max.Min.norm(W) to compute the distance matrix.
c.  Integrated matrix obtained using Similarity Network Fusion. Use dist \<- 1 - NetPreProc::Max.Min.norm(W) to compute the distance matrix.
d.  [GROUP] Integrated matrix obtained using NEMO. Use dist \<- 1 - NetPreProc::Max.Min.norm(W) to compute the distance matrix.

```{r}
#Get the number of clusters for PAM equal to the number of unique subtypes
n_of_clusters <- length(unique(subtypes$Subtype_Selected))

PAM.compute <- function(similarity_matrix, k) {
  #We obtain the normalized distance matrix form the similarity matrix of the i-th assay
  distances <- as.dist(1 - NetPreProc::Max.Min.norm(similarity_matrix))
  #We compute and return the PAM Clustering for the i-th assay
  pam(distances, k=n_of_clusters)
}

# PAM clustering for each assay
pam_results <- list();
pam_next_count <- 1;
for(i in 1:length(W_list)) {
  pam_results[[pam_next_count]] <- PAM.compute(W_list[[i]], n_of_clusters)
  pam_next_count <- pam_next_count + 1;
}

# PAM clustering for average matrix of the assays
pam_results[[pam_next_count]] <- PAM.compute(W_averages, n_of_clusters)
pam_next_count <- pam_next_count + 1;

  # PAM clustering of the SNF integration matrix
pam_results[[pam_next_count]] <- PAM.compute(W_int, n_of_clusters)
pam_next_count <- pam_next_count + 1;

#TODO: [Group] PAM clustering
```

9.  [GROUP] NEMO provides the possibility of performing clustering using another approach called Spectral Clustering [6]. Use the function nemo.clustering() to test this approach.

```{r}

```

10. [OPTIONAL] Apply Spectral Clustering on the integrated matrix obtained using Similarity Network Fusion (an implementation of spectral clustering is SNFtool::spectralClustering(), which is the same exploited in nemo.clustering()).

```{r}
SNF_spectral_clustering <- SNFtool::spectralClustering(W_int, K=n_of_clusters)
```

11. Compare the clusterings obtained by each considered approach w.r.t. the iCluster disease subtypes. Make tables and plots to show the results and discuss them.

```{r}

```